src/cli/cli_otns.cpp

```diff
@@ -0,0 +1,283 @@
+
+    // now, send the coap message
+    SuccessOrExit(error = otThreadGetLeaderAloc(mInterpreter.mInstance, &coapDestinationIp));
+
+    message = otCoapNewMessage(mInterpreter.mInstance, NULL);
+    VerifyOrExit(message != NULL, error = OT_ERROR_NO_BUFS);
+
+    otCoapMessageInit(message, coapType, coapCode);
+    otCoapMessageGenerateToken(message, ot::Coap::Message::kDefaultTokenLength);
+    SuccessOrExit(error = otCoapMessageAppendUriPathOptions(message, OTNS_URI_PATH_COLLECTION_TREE));
+
+    SuccessOrExit(error = otCoapMessageSetPayloadMarker(message));
+
+    assert(otns.mCollectionTreeSensorId > 0);
+    senseData.sensorId  = HostSwap32(otns.mCollectionTreeSensorId);
+    senseData.milliTime = HostSwap32(otPlatAlarmMilliGetNow());
+    senseData.senseVal  = HostSwap32(0xAABBCCDD);
+
+    SuccessOrExit(error = otMessageAppend(message, &senseData, sizeof(senseData)));
+
+    memset(&messageInfo, 0, sizeof(messageInfo));
+    messageInfo.mPeerAddr = coapDestinationIp;
+    messageInfo.mPeerPort = OT_DEFAULT_COAP_PORT;
+
+    error = otCoapSendRequest(mInterpreter.mInstance, message, &messageInfo, NULL, &otns);
+
+exit:
+
+    if ((error != OT_ERROR_NONE) && (message != NULL))
+    {
+        otMessageFree(message);
+    }
+
+    return;
+}
+
+void OTNS::HandleCollectionTreeRequest(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    static_cast<OTNS *>(aContext)->HandleCollectionTreeRequest(aMessage, aMessageInfo);
+}
+
+void OTNS::HandleCollectionTreeRequest(otMessage *aMessage, const otMessageInfo *aMessageInfo)
+{
+    otError    error           = OT_ERROR_NONE;
+    otMessage *responseMessage = NULL;
+    otCoapCode responseCode    = OT_COAP_CODE_EMPTY;
+    char       responseContent = '0';
+    SenseData  senseData;
+
+    if (otThreadGetDeviceRole(mInterpreter.mInstance) != OT_DEVICE_ROLE_LEADER)
+    {
+        otLogWarnCoap("this is not the leader");
+        ExitNow(error = OT_ERROR_INVALID_STATE);
+    }
+
+    assert(otCoapMessageGetType(aMessage) == OT_COAP_TYPE_CONFIRMABLE);
+
+    assert(otMessageGetOffset(aMessage) + sizeof(SenseData) == otMessageGetLength(aMessage));
+    VerifyOrExit(otMessageRead(aMessage, otMessageGetOffset(aMessage), &senseData, sizeof(SenseData)) ==
+                 sizeof(SenseData));
+
+    OtnsStatusPush("ctss=%lx:%lx:%lx", HostSwap32(senseData.sensorId), HostSwap32(senseData.milliTime),
+                   HostSwap32(senseData.senseVal));
+
+    responseCode = OT_COAP_CODE_VALID;
+
+    responseMessage = otCoapNewMessage(mInterpreter.mInstance, NULL);
+    VerifyOrExit(responseMessage != NULL, error = OT_ERROR_NO_BUFS);
+
+    otCoapMessageInitResponse(responseMessage, aMessage, OT_COAP_TYPE_ACKNOWLEDGMENT, responseCode);
+    SuccessOrExit(error = otCoapMessageSetToken(responseMessage, otCoapMessageGetToken(aMessage),
+                                                otCoapMessageGetTokenLength(aMessage)));
+
+    if (otCoapMessageGetCode(aMessage) == OT_COAP_CODE_GET)
```

Comment:
```suggestion
    VerifyOrExit(otCoapMessageGetCode(aMessage) == OT_COAP_CODE_POST, error == OT_ERROR_PARSE);
```
And move this check to the beginning of this function?